### ISSUE TITLE
Forcing generics to have the same keys confuses Typescript

### ISSUE THREAD

**Issue by rubenlg** [Microsoft/TypeScript#17150]
**TypeScript Version:**  2.4.0

**Code**

```ts
// An application object with two generics. The meaning of these doesn't matter here.
interface AppObject<State, Proxy> {
    // ...
}

// A dictionary of AppObject instances
export type AppObjectDict<
  State, 
  Proxy extends {[K in keyof State]: Proxy[K]}> = {
    [K in keyof State]: AppObject<State[K], Proxy[K]>
  };

// Produces a dictionary of proxies
function getProxy<
  State, Proxy extends {[K in keyof State]: Proxy[K]}>(
  val: AppObjectDict<State, Proxy>): Proxy {
  throw Error('Not implemented');
}

// Produces a dictionary of states
function getState<
  State, Proxy extends {[K in keyof State]: Proxy[K]}>(
  val: AppObjectDict<State, Proxy>): State {
  throw Error('Not implemented');
}

interface StateA { x: number; }
interface ProxyA { a: string; }

interface StateB { y: string; }
interface ProxyB { b: number; }

function test(x: AppObject<StateA, ProxyA>, y: AppObject<StateB, ProxyB>) {
  const state = getState({obj1: x, obj2: y});
  const obj1s = state.obj1;

  const proxy = getProxy({obj1: x, obj2: y});
  const obj1p = proxy.obj1;
}
```

**Expected behavior:**

Both obj1s and obj1p are correctly typed as StateA and ProxyA respectively.

**Actual behavior:**

The compiler gives up on obj1p, typing it as any.

**Suggestion 1**
This way of specifying that State and Proxy are meant to have the same keys in AppObjectDict is a bit awkward. It would be great to have some other language construct to describe that situation. Perhaps something like this:

```ts
export type AppObjectDict<Keys, State haskeys Keys, Proxy haskeys Keys> = {
    [K in Keys]: AppObject<State[K], Proxy[K]>
  };
```

**Suggestion 2**
If you don't want to increase the languate syntax, perhaps the compiler could support this instead:
```ts
export type AppObjectDict<State, Proxy> = {
    [K in (keyof State & keyof Proxy)]: AppObject<State[K], Proxy[K]>
  };
```
If you plug this in the example above, the compiler decides that State and Proxy don't have any key in common, and they are always `{}`.

**Comment by olegdunkan** [comment]
> If you don't want to increase the languate syntax, perhaps the compiler could support this instead:

```ts
export type AppObjectDict<State, Proxy> = {
    [K in (keyof State & keyof Proxy)]: AppObject<State[K], Proxy[K]>
  };
```
you can write this 


```ts
type Common<T, U> = {[K in keyof (T | U)]: T[K] | U[K]};

var z: Common<{ a: number }, { a: string, b: boolean }>['a'] //type of z is number | string

export type AppObjectDict<State, Proxy> = {
    [K in keyof (State | Proxy)]: AppObject<State[K], Proxy[K]>
  };
```

**Comment by rubenlg** [comment]
Thanks @olegdunkan!

`(keyof State & keyof Proxy)` is funcionally equivalent to `keyof (State|Proxy)`, but the compiler provides a cleaner type for the latter. They can be used in the same situation because the former has just a bunch of invalid types on top of the latter. Still, your type is cleaner and should be the one properly supported by the compiler.

But in fact, if you plug that version of `AppStateDict` in the example above, it doesn't compile because same as with the former type, the compiler drops all the keys after calling to `getProxy()` or `getState()`, so `state.obj1` and `proxy.obj1` are both invalid expressions.

Here is the full code with your proposal, for easy copy&pasting:

```ts
// An application object with two generics. The meaning of these doesn't matter here.
interface AppObject<State, Proxy> {
    // ...
}

export type AppObjectDict<State, Proxy> = {
    [K in keyof (State | Proxy)]: AppObject<State[K], Proxy[K]>
  };

// Produces a dictionary of proxies
function getProxy<State, Proxy>(val: AppObjectDict<State, Proxy>): Proxy {
  throw Error('Not implemented');
}

// Produces a dictionary of states
function getState<State, Proxy>(val: AppObjectDict<State, Proxy>): State {
  throw Error('Not implemented');
}

interface StateA { x: number; }
interface ProxyA { a: string; }

interface StateB { y: string; }
interface ProxyB { b: number; }

function test(x: AppObject<StateA, ProxyA>, y: AppObject<StateB, ProxyB>) {
  const state = getState({obj1: x, obj2: y});
  const obj1s = state.obj1;

  const proxy = getProxy({obj1: x, obj2: y});
  const obj1p = proxy.obj1;
}
```

As I mentioned, this doesn't compile because State and Proxy are inferred to be `{}` on the calls to `getState()` and `getProxy()`.

**Comment by rubenlg** [comment]
This is still an issue in 2.8. I simplified the example a lot and made it a bit easier to understand by using the concept of a "pair". I ran it through TS 2.8, and the compiler still gives up typing things. The problem seems to remain the same, the compiler gets confused by the idea of the two generics sharing the same keys, and seems to compute an empty intersection of keys:

```typescript
interface Pair<A, B> {
  a: A;
  b: B;
}

type PairDict<ADict, BDict> = {
  [K in keyof(ADict|BDict)]: Pair<ADict[K], BDict[K]>;
};

// Identity function, just to force type checking/inference.
function getPairDict<ADict, BDict>(data: PairDict<ADict, BDict>):
    PairDict<ADict, BDict> {
  return data;
}

const wrapped = getPairDict({
  first: {a: 5, b: 'five'},
  second: {a: 7, b: 'seven'},
});

// Error, property first does not exist on type PairDict<{}, {}>.
console.log(wrapped.first.a);
```

**Comment by mhegazy** [comment]
Inference on mapped types only works for [homomorphic mapped types](https://www.typescriptlang.org/docs/handbook/advanced-types.html#mapped-types). these are the only ones the compiler can reason about the reverse mapping.. `PairDict` is not a homomorphic mapped type since it iterates on the keys of two types. 

I think the definition of `getPairDict` is unnecessarily complicated, i think it should be:

```ts
// Identity function, just to force type checking/inference.
function getPairDict<T extends PairDict<any, any>>(data:T):
    T {
    return data;
}
```

**Comment by rubenlg** [comment]
Thanks Mohamed  for the explanation about homomorphic mapped types.

Your proposal for getPairDict unfortunately looses ADict and BDict. My real use case is not a function, but a class, with methods that should return and/or operate on ADict or BDict, so I need to keep them.

A workaround that preserves the correct types if anyone else encounters this problem is as follows:

```typescript
type PairDict<ADict, BDict> = {
  [K in keyof ADict]: Pair<ADict[K], {}>;
} & {
  [K in keyof BDict]: Pair<{}, BDict[K]>;
};
```

Is this a hack that the compiler will not accept in future releases, or is it safe?

**Comment by mhegazy** [comment]
You can use ‘T[“a”]’ to get access to the type of ADict

**Comment by rubenlg** [comment]
`T` doesn't have any property named `a`. In the example above, `a` and `b` are properties of the objects under each key of `T`. I could use `Record<keyof T, T[keyof T]["a"]>`, but that only works because in this made-up toy example, I happen to know a property of the thing called `Pair` that carries directly the type I need.

In a real-world example where the equivalent of `Pair` is a more complex interface where `A` and/or `B` are not properties but arguments of methods, it's not that simple to extract these types.

My real-world equivalent of `Pair` can look more like this (in the line of what I wrote in the original comment):

```typescript
export interface AppObject<State, Proxy> {
  render(state: State): void;
  bind(store: Store<State>): Proxy;
}
```

The equivalent of `PairDict` would be an `AppObjectDict`, and from users of that I could extract Proxy using 2.8's `ReturnType<>` on the `bind` method, but I could not extract State from the render method.

Anyway, I think this can be closed, since you explained clearly why the compiler can't support directly what I requested (creating a mapped type with the keys of two objects).
